File: scripts/core/pricing.py
import json
from pathlib import Path
import math

RULES_PATH = Path("data/pricing_rules.json")


def load_rules():
    with open(RULES_PATH, "r", encoding="utf-8-sig") as f:
        return json.load(f)


def calculate_fee(rule, amount):
    if not rule:
        return 0

    if "fee" in rule:
        return rule["fee"]

    if rule.get("type") == "fixed":
        return rule.get("value", 0)

    if rule.get("type") == "tiered" or "tiers" in rule:
        for tier in rule.get("tiers", []):
            if amount <= tier["max"]:
                return tier["fee"]

    return 0


def calculate_total_cost(
    company_name,
    target_npr,
    rate,
    deposit_method="bank_transfer"
):
    rules = load_rules()
    company_rules = rules.get(company_name, {})

    remittance_amount = math.ceil(target_npr / rate)

    service_fee = calculate_fee(
        company_rules.get("service_fee"),
        remittance_amount
    )

    deposit_rule = company_rules.get("deposit_methods", {}).get(
        deposit_method,
        {}
    )

    deposit_fee = calculate_fee(deposit_rule, remittance_amount)

    total_jpy = remittance_amount + service_fee + deposit_fee

    return {
        "remittance_amount": remittance_amount,
        "service_fee": service_fee,
        "deposit_fee": deposit_fee,
        "total_jpy": total_jpy
    }

File: js/pricing.js
export function getDefaultFee(rule, amount = 100000) {
  if (!rule) return 0;
  if ("fee" in rule) return Number(rule.fee);
  if (rule.type === "fixed") return Number(rule.value || 0);
  if (rule.type === "tiered" || rule.tiers) {
    for (const tier of rule.tiers || []) {
      if (amount <= Number(tier.max)) return Number(tier.fee);
    }
  }
  return 0;
}

export function getDynamicServiceFee(company, targetNpr, pricingRules) {
  const rate = Number(company.rate || 0);
  if (!rate || !targetNpr) return 0;
  const remittanceAmount = Math.ceil(targetNpr / rate);
  const rules = pricingRules[company.company_name] || {};
  return getDefaultFee(rules.service_fee, remittanceAmount);
}

export function calculateCompanyCost(company, targetNpr, pricingRules, userFees) {
  const rate = Number(company.rate || 0);
  if (!rate || !targetNpr) {
    return { ...company, remittance_amount: 0, service_fee: 0, deposit_fee: 0, required_jpy: 0 };
  }

  const remittanceAmount = Math.ceil(targetNpr / rate);
  const feeState = userFees[company.company_name] || {};
  const serviceFee = feeState.service_fee_manually_changed
    ? Number(feeState.service_fee || 0)
    : getDynamicServiceFee(company, targetNpr, pricingRules);
  const depositFee = Number(feeState.deposit_fee || 0);

  return {
    ...company,
    remittance_amount: remittanceAmount,
    service_fee: serviceFee,
    deposit_fee: depositFee,
    required_jpy: remittanceAmount + serviceFee + depositFee
  };
}
